File: ports/java/src/core/graphics/BcsPainter.java
package core.graphics;

import java.util.concurrent.locks.ReentrantLock;
import core.kernel.BcsObject;

public class BcsPainter extends BcsObject {
    private BcsCanvas canvas;
    private boolean active;
    private final ReentrantLock lock = new ReentrantLock();

    public BcsPainter(BcsObject parent) {
        super(parent);
        this.active = false;
    }

    public boolean begin(BcsCanvas canvas) {
        lock.lock();
        try {
            if (active) return false;
            this.canvas = canvas;
            this.active = true;
            return true;
        } finally {
            lock.unlock();
        }
    }

    public void end() {
        lock.lock();
        try {
            this.active = false;
            this.canvas = null;
        } finally {
            lock.unlock();
        }
    }
}

File: ports/csharp/src/core/graphics/BcsPainter.cs
using System;
using Bcs.Core.Kernel;

namespace Bcs.Core.Graphics
{
    public class BcsPainter : BcsObject
    {
        private BcsCanvas _canvas;
        private bool _active;
        private readonly object _lock = new object();

        public BcsPainter(BcsObject parent = null) : base(parent)
        {
            _active = false;
        }

        public bool Begin(BcsCanvas canvas)
        {
            lock (_lock)
            {
                if (_active) return false;
                _canvas = canvas;
                _active = true;
                return true;
            }
        }

        public void End()
        {
            lock (_lock)
            {
                _active = false;
                _canvas = null;
            }
        }
    }
}
